problem/0000-ci-health-metrics.md

```diff
@@ -0,0 +1,81 @@
+- Contribution Name: ci-health-metrics
+- Implementation Owner: @alpeb, @kleimkuhler
+- Start Date: 2020-03-16
+- Target Date: 2020-04-03
+- RFC PR: [linkerd/rfc#11](https://github.com/linkerd/rfc/pull/11)
+- Linkerd Issue:
+  [linkerd/linkerd2#4176](https://github.com/linkerd/linkerd2/issues/4176)
+- Reviewers: @grampelberg
+
+# Summary
+
+[summary]: #summary
+
+This RFC proposes a service that would monitor the results from each CI run. It
+should allow us to catch consistently flaky tests and have an overview of the CI
+in terms of success rates and time performance.
+
+# Problem Statement (Step 1)
+
+[problem-statement]: #problem-statement
+
+CI jobs inevitably fail occasionally, either because of real problems with the
+code, or because of tests flakiness. We'd like to be able to be informed about
+what tests are being persistently flaky, without having to dig into the CI run
+results themselves. This will allow us to properly prioritize work to fix those
+failures.
+
+Besides identifying the most common failures, by surfacing error metrics from the
+CI jobs we'll also have at our disposal an overview of CI's performance, in
+terms of things like success rates and time taken by the jobs. This information
+can be aggregated at the global, workflow and jobs levels.
+
+## Phase 1
```

Comment:
I might be jumping to implementation details too soon, but have you considered junit output instead of something github specific?

Reply:
I'm also curious about this. I know with GH workflows we have already tied CI to GH specific functionality, but is there a more general output we can use so that we try to remain somewhat independent from where our integration tests are running?

Reply:
Unfortunately, currently GH Actions don't allow exporting test result data, like other tools do in xUnit format or what have you. Besides their own UI the only way to surface that data is through:

1) Annotations, that are then accessible through the Workflows and Checks API. This requires that our tests output annotation strings, which are formatted like `::error file=app.js,line=10,col=15::MessageHere`
2) Have our tests output any format (like xUnit reports), that we can store as a GH artifact.

My initial plan (not outlined in this Step 1 of the RFC), was to do (1) above and then have a Prometheus export consume that and then have Grafana show it. I showed a sample end result during the demo last Friday:
![image](https://user-images.githubusercontent.com/554287/76978769-21244700-6905-11ea-8d28-0a4127f09e41.png)
(Note the data backing that dashboard is all fake, I haven't hooked any of this yet)

As for the xUnit route, we could pull that from the artifacts, but I haven't been able to find any compelling tool that could leverage that, specifically for showing historical trends (or persistently flaky tests), which I believe is what we're after. I did find [UnitTH](http://junitth.sourceforge.net/) but that looks pretty dated, and there's also [allure2](https://github.com/allure-framework/allure2), which seems would require more effort to integrate. Do you have something else in mind?